src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,319 @@
+    }
+    """
+    res_dict = {}
+    res_dict["upa"] = "{6}/{0}/{4}".format(*obj_info)
+    res_dict["name"] = obj_info[1]
+    res_dict["type"] = obj_info[2]
+    res_dict["timestamp"] = obj_info[3]
+    return res_dict
+
+
+def list_objects(wsid, conf, batch_size, filter_objects_name_by=None):
+    """
+    List all objects information given a workspace ID
+    """
+
+    maxObjectID = conf.ws.get_workspace_info({"id": wsid})[4]
+    batch_input = [
+        [idx + 1, idx + batch_size] for idx in range(0, maxObjectID, batch_size)
+    ]
+    objs = [
+        conf.ws.list_objects(_list_objects_params(wsid, min_id, max_id))
+        for min_id, max_id in batch_input
+    ]
+    res_objs = list(itertools.chain.from_iterable(objs))
+    if filter_objects_name_by:
+        res_objs = [
+            obj for obj in res_objs if obj[2].startswith(filter_objects_name_by)
+        ]
+    return res_objs
+
+
+def process_input(conf):
+    """
+    Download .fa and .meta files from workspace and save a copy under output_dir
+    """
+    while True:
+        upa, obj_info = conf.queue.get(block=True)
+        if not upa:
+            print("Stopping")
+            break
+        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+        # upa file is downloaded to cfn
+        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+        # each upa in output_dir as a seperate directory
+        dstd = os.path.join(conf.pth, upa)
+        os.makedirs(dstd, exist_ok=True)
+
+        dst = os.path.join(dstd, f"{upa}.fa")
+        # copy downloaded upa file to output_dir. Hard link might cause invalid cross-device link problem
+        shutil.copy(cfn, dst)
+
+        metafile = os.path.join(dstd, f"{upa}.meta")
+        # save meta file with relevant object_info
+        json.dump(_process_object_info(obj_info), open(metafile, "w"), indent=2)
+
+        print("Completed %s" % (upa))
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="PROTOTYPE - Download genome files from the workspace service (WSS)."
+    )
+
+    required = parser.add_argument_group("required named arguments")
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+    required.add_argument(
+        "--project_dir",
+        type=str,
+        help="Path points to Collections repository",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--output_dir",
+        type=str,
+        help="Output directoy to save genome objects.",
+    )
+    optional.add_argument(
+        "--job_dir",
+        type=str,
+        help="Job directoy of source link",
+    )
```

Comment:
I am not sure we need all those dirs. `root_dir` and `project_dir` seems to be the same thing. `output_dir` and `job_dir` can also be built based on `root_dir`. I also think we could completely get rid of `job_dir`.

Reply:
For `job_dir` see the comments above, but it also seems to me that we only need `root_dir` unless there's something I'm not understanding

Reply:
project_dir, output_dir, and job_dir are removed

Reply:
:+1: